mesonbuild/dependencies/base.py

```diff
@@ -2137,6 +2140,79 @@ def log_tried(self):
         return 'framework'
 
 
+class DependencyFactory:
+
+    """Factory to get dependencies from multiple sources.
+
+    This class provides a constructor that takes a set of names and classes
+    for various kinds of dependencies. This simplifies getting a single
+    dependency from multiple sources.
+
+    :name: The name of the dependency. This will be passed as the name
+        parameter of the each dependency unless it is overridden on a per
+        type basis.
+    :methods: An ordered list of DependencyMethods. This is the order
+        dependencies will be returned in unless they are removed by the
+        _process_method function
+    :*_name: This will overwrite the name passed to the coresponding class.
+        For example, if the name is 'zlib', but cmake calls the dependency
+        'Z', then using `cmake_name='Z'` will pass the name as 'Z' to cmake.
+    :*_class: A *type* or callable that creates a class, and has the
+        signature of an ExternalDependency
+    :system_class: If you pass DependencyMethods.SYSTEM in methods, you must
+        set this argument.
+    """
+
+    def __init__(self, name: str, methods: T.List[DependencyMethods], *,
+                 pkgconfig_name: T.Optional[str] = None,
+                 pkgconfig_class: T.Type[PkgConfigDependency] = PkgConfigDependency,
+                 cmake_name: T.Optional[str] = None,
+                 cmake_class: T.Type[CMakeDependency] = CMakeDependency,
+                 configtool_class: T.Optional[T.Type[ConfigToolDependency]] = None,
+                 framework_name: T.Optional[str] = None,
+                 framework_class: T.Type[ExtraFrameworkDependency] = ExtraFrameworkDependency,
+                 system_class: T.Type[ExternalDependency] = ExternalDependency):
+
+        if DependencyMethods.CONFIG_TOOL in methods and not configtool_class:
+            raise DependencyException('A configtool must have a custom class')
+
+        self.methods = methods
+        self.classes = {
+            # Just attach the correct name right now, either the generic name
+            # or the method specific name.
+            DependencyMethods.EXTRAFRAMEWORK: functools.partial(framework_class, framework_name or name),
+            DependencyMethods.PKGCONFIG: functools.partial(pkgconfig_class, pkgconfig_name or name),
+            DependencyMethods.CMAKE: functools.partial(cmake_class, cmake_name or name),
+            DependencyMethods.SYSTEM: functools.partial(system_class, name),
+            DependencyMethods.CONFIG_TOOL: None,
```

Comment:
Maybe this could be written as iterating over the non-obsolete dependency methods, rather than hardcoding a list? Unclear if that would be worthwhile, or how often that list might change...

Reply:
I'm not sure if it can, or if the resulting code would really be any more maintainable, given that the methods are invoked differently.